virtual_ecosystem/core/core_components.py

```diff
@@ -218,33 +272,120 @@ def __post_init__(self, config: Config) -> None:
         )
 
+        # Record the number of layers and layer indices
         self.n_layers = len(self.layer_roles)
+        self.layer_indices = np.arange(0, self.n_layers)
+
+        # Set up the various indices onto those roles
+        self.role_indices_bool = {
+            layer_role: self.layer_roles == layer_role
+            for layer_role in ("above", "canopy", "surface", "topsoil", "subsoil")
+        }
+        self.role_indices = {
+            ky: np.nonzero(vl)[0] for ky, vl in self.role_indices_bool.items()
+        }
+
+        # Add the `all_soil` and `atmospheric` indices
+        self.role_indices["all_soil"] = np.concatenate(
+            [self.role_indices["topsoil"], self.role_indices["subsoil"]]
+        )
+        self.role_indices_bool["all_soil"] = np.logical_or(
+            self.role_indices_bool["topsoil"], self.role_indices_bool["subsoil"]
+        )
+        self.role_indices_bool["atmosphere"] = np.logical_not(
+            self.role_indices_bool["all_soil"]
+        )
+        self.role_indices["atmosphere"] = np.where(
+            self.role_indices_bool["atmosphere"]
+        )[0]
+
+        # Check that the maximum depth of the last layer is greater than the max depth
+        # of microbial activity.
+        if self.soil_layers[-1] > -self.max_depth_of_microbial_activity:
+            to_raise = ConfigurationError(
+                "Maximum depth of soil layers is less than the maximum depth "
+                "of microbial activity"
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        # Generate indices and layer information for the soil layers containing
+        # significant microbial activity
+        soil_layer_boundaries = np.array([0, *self.soil_layers])
+        self.soil_layer_thickness = -np.diff(soil_layer_boundaries)
+        self.soil_layer_active_thickness = np.clip(
+            np.minimum(
+                self.soil_layer_thickness,
+                (soil_layer_boundaries + self.max_depth_of_microbial_activity)[:-1],
+            ),
+            a_min=0,
+            a_max=np.inf,
+        )
+
+        self.role_indices["active_soil"] = self.role_indices["all_soil"][
+            self.soil_layer_active_thickness > 0
+        ]
+        self.role_indices_bool["active_soil"] = np.array(
+            [v in self.role_indices["active_soil"] for v in np.arange(self.n_layers)]
+        )
+
+        # Create a private template data array with the simulation structure. This
+        # should not be accessed directly to avoid the chance of someone modifying the
+        # actual template.
+        self._array_template: DataArray = DataArray(
+            np.full((self.n_layers, self.n_cells), np.nan),
+            dims=("layers", "cell_id"),
+            coords={
+                "layers": self.layer_indices,
+                "layer_roles": ("layers", self.layer_roles),
+                "cell_id": np.arange(self.n_cells),
+            },
+        )
```

Comment:
Given that you need to deepcopy this array, I don't see much point on keeping a private copy of it. I guess there might be some performance improvement but given that all of these arrays will be created when setting up the simulation and not when running it (I think), the overall gains will be minimal - if any. I would just create this in the `from_template` below, which also avoids risking careless users modifying the template at runtime (that's the paranoid's vision).

Reply:
That's an interesting point, @dalonsoa and something I had parked in the back of my head. I think the different models differ in whether they update an existing DataArray or simply replace it with an updated one. I'm sure we have `from_template()` being used in model `update` methods, so it will be used when running. That _might_ be something we want to avoid doing, or it might be that the `deepcopy` is worth hanging on to.

I think we might want to start leaving a few comments in the code base like `# PERFORMANCE - does deepcopy of the template here provide any real benefit over on the fly creation".